### PR TITLE
refactor: python typing improvements

### PR DESCRIPTION
**Pull request by another-rex**
A series of typing improvements I made while working on the cursor changes. No logic or behavior changes have been made, so should perform identically with the existing deployed code.

Changes:
- Add mypy-protobuf extension, which enables the `--mypy_out=` flag for protoc. This generates a typing stub for the grpc protobuf output, which can be used by different python type checking tools/extensions. (does not have to be mypy)
- Generate and commit the pyi files, and update the readme / script to help generate them in the future.
- Move helper functions in models.py to their own file (model_helpers.py)
- Add explicit type definitions to models in `models.py`
  - These types are not technically "correct", but are the types that will get returned by datastore. This allows us to more easily label the responses from ndb queries. 
  - Added an exclusion to the assignment type check of models.py to the pyproject.toml for pyright (used by vscode and pylance). If you use a different type checker, you might want to look into ignoring the errors in models.py in your own config.

Sadly yield still causes problems typing wise, so turning on "strict" typing is still not possible, details for anyone interested in why:

> Using the yield anywhere in a function turns the function into a generator,
> The left hand side output of yield is received when the caller calls send() on the returned generator.
> Python assumes that the sent type will always be the same throughout a single function, so the send type is defined at the top of the function.
> ndb uses yield as its own custom async implementation, the left hand side output of yield is the result of the future returned from the right hand side of yield.
> This means the output of a yield call will depend on the right hand side of the yield, which is impossible to represent in python's type hinting system.

### CODE REVIEW

File: osv/model_helpers.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Datastore types."""
+
+import datetime
+
+from typing import List, Self
```
Comment (andrewpollock) [inline osv/model_helpers.py:18]: pylint seems to feel these is unnecessary...

**Review by andrewpollock — APPROVED** [review]
Awesome code quality improvement, thank you!

File: osv/model_helpers.py (hunk 1 above)
Comment (another-rex) [reply osv/model_helpers.py]: Ah yep, fixed! I really should set up some pre-push linting check on my machine.

File: osv/models.py [hunk 2]
```diff
@@ -13,14 +13,14 @@
...
-import datetime
+from datetime import datetime
 import enum
 import logging
 import re
 import os
 
 from urllib.parse import urlparse
-from typing import List
+from typing import List, Self
```
Comment (michaelkedar) [inline osv/models.py:23]: is `List` still used in this file?

**Review by michaelkedar — APPROVED** [review]
LGTM, a few very small things

File: gcp/api/server.py [hunk 3]
```diff
@@ -25,6 +25,7 @@
 import os
 import threading
 import time
+import concurrent.futures
```
[the remark this answers is not in the document]
Comment (another-rex) [reply gcp/api/server.py:28]: I think it's used on line 1385 to start the server: 
```
  server = grpc.server(concurrent.futures.ThreadPoolExecutor(max_workers=5))
```

[the remark this answers is not in the document]
Comment (another-rex) [reply gcp/api/server.py:28]: But concurrent itself is not used, I'll remove that.

File: osv/models.py [hunk 4]
```diff
@@ -335,7 +342,7 @@ def repo_url(self):
     return None
 
   @classmethod
-  def get_by_id(cls, vuln_id, *args, **kwargs):
+  def get_by_id(cls, vuln_id, *args, **kwargs) -> Self:
```
[the remark this answers is not in the document]
Comment (another-rex) [reply osv/models.py:345]: Good find!

File: gcp/api/server.py [hunk 5]
```diff
@@ -759,7 +765,7 @@ def to_response(b):
   else:
     context.service_context.abort(grpc.StatusCode.INVALID_ARGUMENT,
                                   'Invalid query.')
-    return None
+    raise ValueError
```
[the remark this answers is not in the document]
Comment (another-rex) [reply gcp/api/server.py:768]: Added a comment to say that this is unreachable, but is there for the type checker to understand that code does not flow past this point.

File: osv/models.py (hunk 2 above)
Comment (another-rex) [reply osv/models.py:23]: Replaced uses of List with list, and removed this import.

**Pull request merged**